### ISSUE TITLE
ERROR [LoggableUncaughtExceptionHandler] An unknown error occurred in an asynchronous event or notification thread

### ISSUE THREAD

**Issue by nibiwodong** [DependencyTrack/dependency-track#1059]
The defect may already be reported! Please search for the defect before creating one.

### Current Behavior:
When I use Jenkins DTrack plugin to upload bom.xml，dtrack api server raise an error: 

`02:57:30.534 INFO [MetricsUpdateTask] Completed metrics update for project: a4cf0b36-b4cf-45ea-b742-9bb7ab909172
02:57:30.535 INFO [MetricsUpdateTask] Executing metrics update for project: 7652e918-e933-4791-9a98-1f6d03407f4f
02:57:33.637 INFO [BomUploadProcessingTask] Processing CycloneDX BOM uploaded to project: f70761b5-b9ae-4d8c-8b34-b6a7d5544d52
02:57:35.703 ERROR [LoggableUncaughtExceptionHandler] An unknown error occurred in an asynchronous event or notification thread
java.lang.NullPointerException: Value in JsonObjects name/value pair cannot be null
	at org.glassfish.json.JsonObjectBuilderImpl.validateValue(JsonObjectBuilderImpl.java:222)
	at org.glassfish.json.JsonObjectBuilderImpl.add(JsonObjectBuilderImpl.java:90)
	at org.dependencytrack.util.NotificationUtil.toJson(NotificationUtil.java:333)
	at org.dependencytrack.notification.publisher.Publisher.prepareTemplate(Publisher.java:84)
	at org.dependencytrack.notification.publisher.AbstractWebhookPublisher.publish(AbstractWebhookPublisher.java:40)
	at org.dependencytrack.notification.publisher.WebhookPublisher.inform(WebhookPublisher.java:32)
	at org.dependencytrack.notification.NotificationRouter.inform(NotificationRouter.java:64)
	at alpine.notification.NotificationService.lambda$alertSubscriber$0(NotificationService.java:100)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
02:57:38.810 INFO [MetricsUpdateTask] Completed metrics update for project: 7652e918-e933-4791-9a98-1f6d03407f4f
02:57:38.811 INFO [MetricsUpdateTask] Executing metrics update for project: 5d15de20-d3e8-4396-9a9c-e14825fbb211
02:57:38.853 INFO [MetricsUpdateTask] Completed metrics update for project: 5d15de20-d3e8-4396-9a9c-e14825fbb211
02:57:38.854 INFO [MetricsUpdateTask] Executing metrics update for project: 61e6c57d-ee4a-4e10-a32b-df7eba935951
02:57:39.273 INFO [BomUploadProcessingTask] Processing CycloneDX dependency graph for project: f70761b5-b9ae-4d8c-8b34-b6a7d5544d52
02:57:39.795 INFO [BomUploadProcessingTask] Processed 352 components and 0 services uploaded to project f70761b5-b9ae-4d8c-8b34-b6a7d5544d52
02:57:39.811 ERROR [LoggableUncaughtExceptionHandler] An unknown error occurred in an asynchronous event or notification thread
java.lang.NullPointerException: Value in JsonObjects name/value pair cannot be null
	at org.glassfish.json.JsonObjectBuilderImpl.validateValue(JsonObjectBuilderImpl.java:222)
	at org.glassfish.json.JsonObjectBuilderImpl.add(JsonObjectBuilderImpl.java:90)
	at org.dependencytrack.util.NotificationUtil.toJson(NotificationUtil.java:333)
	at org.dependencytrack.notification.publisher.Publisher.prepareTemplate(Publisher.java:84)
	at org.dependencytrack.notification.publisher.AbstractWebhookPublisher.publish(AbstractWebhookPublisher.java:40)
	at org.dependencytrack.notification.publisher.WebhookPublisher.inform(WebhookPublisher.java:32)
	at org.dependencytrack.notification.NotificationRouter.inform(NotificationRouter.java:64)
	at alpine.notification.NotificationService.lambda$alertSubscriber$0(NotificationService.java:100)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
02:58:39.542 ERROR [VulnerabilityAnalysisTask] An unexpected error occurred performing a vulnerability analysis task
javax.jdo.JDOObjectNotFoundException: No such database row
	at org.datanucleus.api.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:563)
	at org.datanucleus.api.jdo.JDOPersistenceManager.jdoRefresh(JDOPersistenceManager.java:502)
	at org.datanucleus.api.jdo.JDOPersistenceManager.refresh(JDOPersistenceManager.java:514)
	at alpine.persistence.AbstractAlpineQueryManager.persist(AbstractAlpineQueryManager.java:401)
	at org.dependencytrack.persistence.QueryManager.addVulnerability(QueryManager.java:1514)
	at org.dependencytrack.tasks.scanners.OssIndexAnalysisTask.processResults(OssIndexAnalysisTask.java:251)
	at org.dependencytrack.tasks.scanners.OssIndexAnalysisTask.analyze(OssIndexAnalysisTask.java:175)
	at org.dependencytrack.tasks.scanners.OssIndexAnalysisTask.inform(OssIndexAnalysisTask.java:112)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.performAnalysis(VulnerabilityAnalysisTask.java:146)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.inform(VulnerabilityAnalysisTask.java:103)
	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:99)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
	Suppressed: javax.jdo.JDODataStoreException: Connection is closed
		at org.datanucleus.api.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:542)
		at org.datanucleus.api.jdo.JDOPersistenceManager.internalClose(JDOPersistenceManager.java:194)
		at org.datanucleus.api.jdo.JDOPersistenceManager.close(JDOPersistenceManager.java:161)
		at alpine.persistence.AbstractAlpineQueryManager.close(AbstractAlpineQueryManager.java:591)
		at org.dependencytrack.tasks.scanners.OssIndexAnalysisTask.processResults(OssIndexAnalysisTask.java:232)
		... 8 common frames omitted
	Caused by: java.sql.SQLException: Connection is closed
		at com.zaxxer.hikari.pool.ProxyConnection$ClosedConnection.lambda$getClosedConnection$0(ProxyConnection.java:493)
		at com.sun.proxy.$Proxy32.clearWarnings(Unknown Source)
		at com.zaxxer.hikari.pool.ProxyConnection.close(ProxyConnection.java:246)
		at org.datanucleus.store.rdbms.ConnectionFactoryImpl$ManagedConnectionImpl.close(ConnectionFactoryImpl.java:571)
		at org.datanucleus.store.connection.ConnectionManagerImpl.closeAllConnections(ConnectionManagerImpl.java:275)
		at org.datanucleus.store.AbstractStoreManager$1.preClose(AbstractStoreManager.java:156)
		at org.datanucleus.ExecutionContextImpl.close(ExecutionContextImpl.java:466)
		at org.datanucleus.api.jdo.JDOPersistenceManager.internalClose(JDOPersistenceManager.java:186)
		... 11 common frames omitted
Caused by: org.datanucleus.exceptions.NucleusObjectNotFoundException: No such database row
	at org.datanucleus.store.rdbms.request.FetchRequest.execute(FetchRequest.java:440)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.fetchObject(RDBMSPersistenceHandler.java:316)
	at org.datanucleus.state.StateManagerImpl.loadFieldsFromDatastore(StateManagerImpl.java:1542)
	at org.datanucleus.state.StateManagerImpl.refreshFieldsInFetchPlan(StateManagerImpl.java:3979)
	at org.datanucleus.api.jdo.state.Hollow.transitionRefresh(Hollow.java:214)
	at org.datanucleus.state.StateManagerImpl.refresh(StateManagerImpl.java:954)
	at org.datanucleus.ExecutionContextImpl.refreshObject(ExecutionContextImpl.java:1608)
	at org.datanucleus.api.jdo.JDOPersistenceManager.jdoRefresh(JDOPersistenceManager.java:497)
	... 12 common frames omitted
02:58:39.551 INFO [PolicyEngine] Evaluating 352 component(s) against applicable policies
02:58:42.420 INFO [PolicyEngine] Policy analysis complete
`

### Environment:
  
  - Dependency-Track Version: 4.2.2
  - Distribution: Docker
  - BOM Format & Version: Use cyclonedx-maven-plugin 2.5.1 create bom.xml
  - Database Server: PostgreSQL
  - Browser: Chrome

### Additional Details:

(e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, gitter, etc)

**Comment by stevespringett** [comment]
Please provide env details. The issue template provided placeholder for this info, but it's not filled in.

**Comment by nscuro** [comment]
> When I use Jenkins DTrack plugin to upload **pom.xml**

Is this a typo or did you indeed upload your Maven project's POM? 
You have to generate a CycloneDX **B**OM first, e.g. with this Maven Plugin: https://github.com/CycloneDX/cyclonedx-maven-plugin

**Comment by stevespringett** [comment]
Per the template:
> ### Environment:
> Dependency-Track Version:
> Distribution: [ Docker | Executable WAR | Traditional WAR ]
> BOM Format & Version:
> Database Server: [ H2 | MSSQL | MySQL | PostgreSQL ]

**Comment by nibiwodong** [comment]
> Per the template:
> 
> > ### Environment:
> > Dependency-Track Version:
> > Distribution: [ Docker | Executable WAR | Traditional WAR ]
> > BOM Format & Version:
> > Database Server: [ H2 | MSSQL | MySQL | PostgreSQL ]

I have update the environment info.

**Comment by nibiwodong** [comment]
> > When I use Jenkins DTrack plugin to upload **pom.xml**
> 
> Is this a typo or did you indeed upload your Maven project's POM?
> You have to generate a CycloneDX **B**OM first, e.g. with this Maven Plugin: https://github.com/CycloneDX/cyclonedx-maven-plugin

Yes, bom.xml. I've got a typo...

**Comment by stevespringett** [comment]
Duplicate of #1026

**Comment by msymons** [comment]
@stevespringett, I do not think that this is a duplicate,  given that #1026 was resolved in v4.3.0 release.

I saw 48 occurrences of the exception overnight on a brand new (not upgraded) 4.3.1 server.  All 48 occurred within the space of 4 minutes and this was after all had been quite happy for 2 hours with DT chuntering away happily processing all the uploads pouring in from Jenkins nightly builds.   It had processed some 75 uploads in the 2 hours before the exceptions occurred.

Is it possible that the problem occurs specifically when DT has kicked off some other background task?  That might explain why things operate fine for a while,  then there is a spasm,  and then things are OK again (because whatever that task was is no longer running).

For 4.3.1

```
00:55:30.618 INFO [InternalAnalysisTask] Starting internal analysis task
00:55:34.851 ERROR [VulnerabilityAnalysisTask] An unexpected error occurred performing a vulnerability analysis task
javax.jdo.JDOObjectNotFoundException: No such database row
	at org.datanucleus.api.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:563)
	at org.datanucleus.api.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1747)
	at alpine.persistence.AbstractAlpineQueryManager.getObjectById(AbstractAlpineQueryManager.java:522)
	at org.dependencytrack.tasks.scanners.InternalAnalysisTask.analyze(InternalAnalysisTask.java:84)
	at org.dependencytrack.tasks.scanners.InternalAnalysisTask.inform(InternalAnalysisTask.java:59)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.performAnalysis(VulnerabilityAnalysisTask.java:165)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.analyzeComponents(VulnerabilityAnalysisTask.java:129)
	at org.dependencytrack.tasks.VulnerabilityAnalysisTask.inform(VulnerabilityAnalysisTask.java:93)
	at alpine.event.framework.BaseEventService.lambda$publish$0(BaseEventService.java:99)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.datanucleus.exceptions.NucleusObjectNotFoundException: No such database row
	at org.datanucleus.store.rdbms.request.FetchRequest.execute(FetchRequest.java:439)
	at org.datanucleus.store.rdbms.RDBMSPersistenceHandler.fetchObject(RDBMSPersistenceHandler.java:316)
	at org.datanucleus.state.StateManagerImpl.loadFieldsFromDatastore(StateManagerImpl.java:1542)
	at org.datanucleus.state.StateManagerImpl.validate(StateManagerImpl.java:5521)
	at org.datanucleus.ExecutionContextImpl.findObject(ExecutionContextImpl.java:3562)
	at org.datanucleus.ExecutionContextImpl.findObject(ExecutionContextImpl.java:3021)
	at org.datanucleus.api.jdo.JDOPersistenceManager.getObjectById(JDOPersistenceManager.java:1742)
	... 10 common frames omitted
00:55:34.855 INFO [NpmAuditAnalysisTask] Starting Node Audit analysis task
```

Database = Azure PostgreSQL 10.16

**Comment by 37b** [comment]
I also had an enormous number of those row errors shortly after upgrading from 4.2.2. I didn't really notice until this week (shame on me).

What alerted me to a problem is numerous users reported getting a token back after uploading a BOM, but not seeing the last upload timestamp update, or new results appear on the UI.

I am also using PostgreSQL (11.10 RDS Aurora)
